File: packages/core-types/src/entity.ts
/**
 * Entity types for Mystira platform.
 * Mirrors Mystira.Shared.Data.Entities in C#.
 */

/**
 * Base entity interface with ID.
 */
export interface Entity {
  /** Unique identifier (ULID recommended) */
  id: string;
}

/**
 * Entity with audit fields.
 */
export interface AuditableEntity extends Entity {
  /** When the entity was created */
  createdAt: string;
  /** Who created the entity */
  createdBy?: string;
  /** When the entity was last updated */
  updatedAt?: string;
  /** Who last updated the entity */
  updatedBy?: string;
}

/**
 * Entity with soft delete support.
 */
export interface SoftDeletableEntity extends AuditableEntity {
  /** Whether the entity is deleted */
  isDeleted: boolean;
  /** When the entity was deleted */
  deletedAt?: string;
  /** Who deleted the entity */
  deletedBy?: string;
}

/**
 * Value object base (immutable, compared by value).
 */
export interface ValueObject {
  /** Value objects should implement equals() */
  equals(other: unknown): boolean;
}

/**
 * Database target for polyglot persistence.
 * Mirrors Mystira.Shared.Data.Polyglot.DatabaseTarget in C#.
 */
export type DatabaseTarget = 'cosmosdb' | 'postgresql' | 'redis';

/**
 * Entity metadata for polyglot routing.
 */
export interface EntityMetadata {
  /** Target database for this entity */
  databaseTarget: DatabaseTarget;
  /** Collection/table name */
  collectionName: string;
  /** Partition key (for Cosmos DB) */
  partitionKey?: string;
}

File: packages/domain/Mystira.Domain/Entities/Entity.cs
namespace Mystira.Domain.Entities;

/// <summary>
/// Base class for entities with a string ID.
/// </summary>
public abstract class Entity : IEntity<string>, IAuditableEntity
{
    /// <summary>
    /// Gets or sets the unique identifier.
    /// </summary>
    public string Id { get; set; } = EntityId.NewId();

    /// <summary>
    /// Gets or sets when the entity was created.
    /// </summary>
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// Gets or sets when the entity was last updated.
    /// </summary>
    public DateTime? UpdatedAt { get; set; }

    /// <summary>
    /// Gets or sets who created the entity.
    /// </summary>
    public string? CreatedBy { get; set; }

    /// <summary>
    /// Gets or sets who last updated the entity.
    /// </summary>
    public string? UpdatedBy { get; set; }
}

/// <summary>
/// Base class for entities with a typed ID.
/// </summary>
/// <typeparam name="TId">The type of the entity identifier.</typeparam>
public abstract class Entity<TId> : IEntity<TId>, IAuditableEntity where TId : notnull
{
    /// <summary>
    /// Gets or sets the unique identifier.
    /// </summary>
    public TId Id { get; set; } = default!;

    /// <summary>
    /// Gets or sets when the entity was created.
    /// </summary>
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// Gets or sets when the entity was last updated.
    /// </summary>
    public DateTime? UpdatedAt { get; set; }

    /// <summary>
    /// Gets or sets who created the entity.
    /// </summary>
    public string? CreatedBy { get; set; }

    /// <summary>
    /// Gets or sets who last updated the entity.
    /// </summary>
    public string? UpdatedBy { get; set; }
}

/// <summary>
/// Base class for soft-deletable entities.
/// </summary>
public abstract class SoftDeletableEntity : Entity, ISoftDeletable
{
    /// <summary>
    /// Gets or sets whether the entity is deleted.
    /// </summary>
    public bool IsDeleted { get; set; }

    /// <summary>
    /// Gets or sets when the entity was deleted.
    /// </summary>
    public DateTime? DeletedAt { get; set; }

    /// <summary>
    /// Gets or sets who deleted the entity.
    /// </summary>
    public string? DeletedBy { get; set; }
}
